core/trino-main/src/main/java/io/trino/operator/ExchangeOperator.java

```diff
@@ -197,6 +215,290 @@ public Page getOutput()
+        }
+    }
+
+    private static class DirectExchangeDataSource
+            implements ExchangeDataSource
+    {
+        private final DirectExchangeClient directExchangeClient;
+
+        private DirectExchangeDataSource(DirectExchangeClient directExchangeClient)
+        {
+            this.directExchangeClient = requireNonNull(directExchangeClient, "directExchangeClient is null");
+        }
+
+        @Override
+        public SerializedPage pollPage()
+        {
+            return directExchangeClient.pollPage();
+        }
+
+        @Override
+        public boolean isFinished()
+        {
+            return directExchangeClient.isFinished();
+        }
+
+        @Override
+        public ListenableFuture<Void> isBlocked()
+        {
+            return directExchangeClient.isBlocked();
+        }
+
+        @Override
+        public void addSplit(RemoteSplit remoteSplit)
+        {
+            DirectExchangeInput taskInput = (DirectExchangeInput) remoteSplit.getExchangeInput();
+            directExchangeClient.addLocation(taskInput.getTaskId(), taskInput.getLocation());
+        }
+
+        @Override
+        public void noMoreSplits()
+        {
+            directExchangeClient.noMoreLocations();
+        }
+
+        @Override
+        public OperatorInfo getInfo()
+        {
+            return directExchangeClient.getStatus();
+        }
+
+        @Override
+        public void close()
+        {
+            directExchangeClient.close();
+        }
+    }
+
+    private static class ExternalExchangeDataSource
+            implements ExchangeDataSource
+    {
+        private final ExchangeSource exchangeSource;
+        private final LocalMemoryContext systemMemoryContext;
+
+        private ExternalExchangeDataSource(ExchangeSource exchangeSource, LocalMemoryContext systemMemoryContext)
+        {
+            this.exchangeSource = requireNonNull(exchangeSource, "exchangeSource is null");
+            this.systemMemoryContext = requireNonNull(systemMemoryContext, "systemMemoryContext is null");
+        }
+
+        @Override
+        public SerializedPage pollPage()
+        {
+            Slice data = exchangeSource.read();
+            systemMemoryContext.setBytes(exchangeSource.getSystemMemoryUsage());
+            if (data == null) {
+                return null;
+            }
+            // TODO: Avoid extra memory copy
+            return readSerializedPage(data.getInput());
+        }
+
+        @Override
+        public boolean isFinished()
+        {
+            return exchangeSource.isFinished();
+        }
+
+        @Override
+        public ListenableFuture<Void> isBlocked()
+        {
+            return asVoid(toListenableFuture(exchangeSource.isBlocked()));
+        }
+
+        @Override
+        public void addSplit(RemoteSplit remoteSplit)
+        {
+            // ignore
+        }
+
+        @Override
+        public void noMoreSplits()
+        {
+            // ignore
```

Comment:
Then it would require another down-casting in the `LazyExchangeDataSource#noMoreSplits`. I added an assertion to the `ExternalExchangeDataSource#addSplit`, so the assumption is that for the `ExternalExchangeDataSource` is is always `no more split`. So it should be okay to have this method as "noop". Added a comment.

Reply:
Yeah - no strong preferences.